src/change_stream/session.rs

```diff
@@ -0,0 +1,137 @@
+/// use futures::stream::TryStreamExt;
+///
+/// let mut cs = coll.watch_with_session(None, None, &mut session).await?;
+/// let results: Vec<_> = cursor.stream(&mut session).try_collect().await?;
+/// #
+/// # Ok(())
+/// # }
+/// ```
+pub struct SessionChangeStream<T>
+where
+    T: DeserializeOwned + Unpin,
+{
+    cursor: SessionCursor<T>,
+    data: ChangeStreamData,
+}
+
+impl<T> SessionChangeStream<T>
+where
+    T: DeserializeOwned + Unpin + Send + Sync,
+{
+    /// Returns the cached resume token that can be used to resume after the most recently returned
+    /// change.
+    ///
+    /// See the documentation
+    /// [here](https://docs.mongodb.com/manual/changeStreams/#change-stream-resume-token) for more
+    /// information on change stream resume tokens.
+    pub fn resume_token(&self) -> Option<&ResumeToken> {
+        todo!()
+    }
+
+    /// Update the type streamed values will be parsed as.
+    pub fn with_type<D: DeserializeOwned + Unpin + Send + Sync>(self) -> SessionChangeStream<D> {
+        todo!()
+    }
+
+    /// Retrieves a [`SessionCursorStream`] to iterate this change stream. The session provided must
+    /// be the same session used to create the cursor.
+    ///
+    /// Note that the borrow checker will not allow the session to be reused in between iterations
+    /// of this stream. In order to do that, either use [`SessionChangeStream::next`] instead or
+    /// drop the stream before using the session.
+    ///
+    /// ```
+    /// # use bson::{doc, Document};
+    /// # use mongodb::{Client, error::Result};
+    /// # fn main() {
+    /// # async {
+    /// # let client = Client::with_uri_str("foo").await?;
+    /// # let coll = client.database("foo").collection::<Document>("bar");
+    /// # let other_coll = coll.clone();
+    /// # let mut session = client.start_session(None).await?;
+    /// #
+    /// use futures::stream::TryStreamExt;
+    ///
+    /// // iterate over the results
+    /// let mut cs = coll.watch_with_session(None, None, &mut session).await?;
+    /// while let Some(doc) = cs.stream(&mut session).try_next().await? {
+    ///     println!("{}", doc);
+    /// }
+    ///
+    /// // collect the results
+    /// let mut cs1 = coll.watch_with_session(None, None, &mut session).await?;
+    /// let v: Vec<Document> = cs1.stream(&mut session).try_collect().await?;
+    ///
+    /// // use session between iterations
+    /// let mut cs2 = coll.watch_with_session(None, None, &mut session).await?;
+    /// loop {
+    ///     let doc = match cs2.stream(&mut session).try_next().await? {
+    ///         Some(d) => d,
+    ///         None => break,
+    ///     };
+    ///     other_coll.insert_one_with_session(doc, None, &mut session).await?;
+    /// }
+    /// # Ok::<(), mongodb::error::Error>(())
+    /// # };
+    /// # }
+    /// ```
+    pub fn stream<'session>(
+        &mut self,
+        session: &'session mut ClientSession,
+    ) -> SessionCursorStream<'_, 'session, T> {
```

Comment:
We may want to introduce a new type here since the change stream's stream will need to be able to resume.

Reply:
Done.  Named it 'SessionChangeStreamValues` to avoid the confusing `StreamStream` suffix it would otherwise have had.  Open to better naming suggestions if inspiration strikes.

Reply:
Yeah I can't really think of a better name either at the moment.